src/auth/downscopedclient.ts

```diff
@@ -0,0 +1,254 @@
+   * When the provided credential expires, a new credential, using the
+   * external account options are retrieved.
+   * Notice DownscopedClient is the broker class mainly used for generate
+   * downscoped access tokens, it is unlikely we call this function in real
+   * use case.
+   * We implement to make this a helper function for testing all cases in getAccessToken().
+   * @param credentials The Credentials object to set on the current client.
+   */
+  setCredentials(credentials: Credentials) {
+    super.setCredentials(credentials);
+    this.cachedDownscopedAccessToken = credentials;
+  }
+
+  async getAccessToken(): Promise<GetAccessTokenResponse> {
+    // If the cached access token is unavailable or expired, force refresh.
+    // The Downscoped access token will be returned in GetAccessTokenResponse format.
+    // If cached access token is unavailable or expired, force refresh.
+    if (
+      !this.cachedDownscopedAccessToken ||
+      this.isExpired(this.cachedDownscopedAccessToken)
+    ) {
+      await this.refreshAccessTokenAsync();
+    }
+    // Return Downscoped access token in GetAccessTokenResponse format.
+    return {
+      token: this.cachedDownscopedAccessToken!.access_token,
+      res: this.cachedDownscopedAccessToken!.res,
+    };
+  }
+
+  /**
+   * The main authentication interface. It takes an optional url which when
+   * present is the endpoint> being accessed, and returns a Promise which
+   * resolves with authorization header fields.
+   *
+   * The result has the form:
+   * { Authorization: 'Bearer <access_token_value>' }
+   */
+  async getRequestHeaders(): Promise<Headers> {
+    throw new Error('Not implemented.');
+  }
+
+  /**
+   * Provides a request implementation with OAuth 2.0 flow. In cases of
+   * HTTP 401 and 403 responses, it automatically asks for a new access token
+   * and replays the unsuccessful request.
+   * @param opts Request options.
+   * @param callback callback.
+   * @return A promise that resolves with the HTTP response when no callback
+   * is provided.
+   */
+  request<T>(opts: GaxiosOptions): GaxiosPromise<T>;
+  request<T>(opts: GaxiosOptions, callback: BodyResponseCallback<T>): void;
+  request<T>(
+    opts: GaxiosOptions,
+    callback?: BodyResponseCallback<T>
+  ): GaxiosPromise<T> | void {
+    throw new Error('Not implemented.');
+  }
+
+  /**
+   * Forces token refresh, even if unexpired tokens are currently cached.
+   * GCP access tokens are retrieved from authclient object/source credential.
+   * Thenm GCP access tokens are exchanged for downscoped access tokens via the
+   * token exchange endpoint.
+   * @return A promise that resolves with the fresh downscoped access token.
+   */
+  protected async refreshAccessTokenAsync(): Promise<CredentialsWithResponse> {
+    // Retrieve GCP access token from source credential.
+    const subjectToken = await (await this.authClient.getAccessToken()).token;
```

Comment:
Why do you need another `await` here? I think this should be sufficient:
`const subjectToken = (await this.authClient.getAccessToken()).token;`